src/wrapped.jl

```diff
@@ -0,0 +1,317 @@
+using Distributions
+
+"""
+    wrapped(d0::UnivariateDistribution, l::Real, u::Real; k::Int=1, kwargs...)
+
+Construct a ``k``-wrapped distribution with support ``[l, u)``.
+
+The resulting density wraps around the interval. With the period ``\\tau = u - l``, the
+density can be expressed as two infinite series in terms of either the density ``f_0`` of
+the unwrapped distribution `d0` or in terms of its characteristic function [`cf`](@ref)
+``\\phi_0``:
+````math
+\\begin{aligned}
+f(x) &= \\frac{1}{\\tau}\\sum_{j=-\\infty}^\\infty f_0\\left(x + \\frac{j}{k}\\tau\\right)\\
+     &= \\frac{1}{\\tau}\\sum_{j=-\\infty}^\\infty
+            \\phi_0\\left(jk\\frac{2\\pi}{\\tau}\\right)
+            \\exp\\left(-ijk\\frac{2\\pi}{\\tau}x\\right)
+\\end{aligned}
+````
+If `k>1`, then the resulting distribution has ``k``-fold symmetry in the interval.
+
+# Keywords
+- `characteristic::Bool`: whether to use the characteristic function to compute the density.
+    If `false`, the density is computed using the logpdf of the unwrapped distribution.
+    If not specified, the `wrapped` method called may choose based on the parameters of
+    `l`, `u`, `tol`, and `d0`.
+- `tol`: target absolute error of `pdf`. May be used to determine whether to use the
+    characteristic function in computations or not.
+
+# Examples
+
+"""
+function wrapped(d::UnivariateDistribution, l::Real, u::Real; kwargs...)
+    return Wrapped(d, promote(l, u)...; kwargs...)
+end
+function wrapped(d::UnivariateDistribution, l::T, u::T; kwargs...) where {T<:Real}
+    return Wrapped(d, l, u; kwargs...)
+end
+
+"""
+    Wrapped
+
+Generic wrapper for a wrapped distribution.
+"""
+struct Wrapped{D<:UnivariateDistribution,S<:ValueSupport,T<:Real} <:
+       UnivariateDistribution{S}
+    unwrapped::D          # the original distribution (untruncated)
+    lower::T              # lower bound
+    upper::T              # upper bound
+    characteristic::Bool  # whether to use characteristic function
+    k::Int                # k-fold symmetry
+    function Wrapped(
+        d::UnivariateDistribution,
+        l::T,
+        u::T;
+        tol::Real=sqrt(eps(float(T))),
+        characteristic::Bool=_wrapped_use_characteristic(d, l, u, tol),
+        k::Int=1,
+    ) where {T<:Real}
+        return new{typeof(d),value_support(typeof(d)),T}(
+            d, l, u, characteristic, k
+        )
+    end
+end
+
+_wrapped_use_characteristic(d::UnivariateDistribution, l, u, tol) = false
+
+minimum(d::Wrapped) = d.lower
+maximum(d::Wrapped) = d.upper - eps(float(d.upper))
```

Comment:
For `Wrapped{D,S,Int}`, this will produce different types for `minimum` and `maximum`, which seems weird to me. In other cases we use the limit, e.g. when it's unbounded, which I think (again naively) would be acceptable here.

Reply:
For a wrapped distribution, `d.lower` and `d.upper` are identified as the same point, so for a discrete distribution, including them both will always give the wrong answer if `d.upper` is in the support of the wrapped distribution. Ideally `maximum(d)` would return the largest point strictly less than `d.upper`, but AFAICT there is no function for this in the Distributions API.

See for example the wrapped Poisson example in https://github.com/JuliaStats/Distributions.jl/issues/1716#issue-1697550141. Because 0 and 15 are the same point, the PMF sums to greater than 1 because `maximum(d)` returned `d.upper`.

A compromise could be returning `d.upper` for continuous distributions and this result for discrete ones.

Reply:
> Ideally `maximum(d)` would return the largest point strictly less than `d.upper`

Distributions defines the support of a `DiscreteUnivariateDistribution` to be `round(Int, minimum(d)):round(Int, maximum(d))`, which means that it doesn't support supports (🙃) with non-integer steps. That implies that the largest point that's strictly less than `d.upper` is `ceil(Int, d.upper) - 1`. Would that work here?

Reply:
> Distributions defines the support of a `DiscreteUnivariateDistribution` to be `round(Int, minimum(d)):round(Int, maximum(d))`, which means that it doesn't support supports (upside_down_face) with non-unit steps.

Oh really? But this doesn't seem to be consistently enforced, e.g.

```julia
julia> d = Binomial(10, 0.5) * 1//2;

julia> support(d)
0//1:1//2:5//1

julia> d = DiscreteNonParametric(randn(5), normalize(rand(5), 1));

julia> support(d)
5-element Vector{Float64}:
 -0.9108476704164078
 -0.5194957770100691
 -0.024410217492521903
  0.15370458561285605
  1.338024968131234
```

Reply:
Hah, amazing. 🤦 I was going based on the generic definition for `DiscreteUnivariateDistribution`, I had forgotten about `LocationScale` and I don't think I even knew `DiscreteNonParametric` existed. To be fair, I did say I was uninformed!